From 603f4deb1e87c819c1830f58c7be4281d981fbbd Mon Sep 17 00:00:00 2001
From: Lenka Doudova <lryznaro@redhat.com>
Date: Mon, 2 Feb 2026 16:46:52 +0100
Subject: [PATCH] Issue 6753 - Port ticket 48896 test

Description:
Port ticket 48896 test into dirsrvtests/tests/suites/password/pwdPolicy_token_test.py

Relates: #6753
Author: Lenka Doudova
Assisted by: Cursor
Reviewer: Mark Reynolds
---
 .../suites/password/pwdPolicy_token_test.py   |  68 ++++++---
 dirsrvtests/tests/tickets/ticket48896_test.py | 139 ------------------
 2 files changed, 44 insertions(+), 163 deletions(-)
 delete mode 100644 dirsrvtests/tests/tickets/ticket48896_test.py

diff --git a/dirsrvtests/tests/suites/password/pwdPolicy_token_test.py b/dirsrvtests/tests/suites/password/pwdPolicy_token_test.py
index ae4eb300f..a3caaa230 100644
--- a/dirsrvtests/tests/suites/password/pwdPolicy_token_test.py
+++ b/dirsrvtests/tests/suites/password/pwdPolicy_token_test.py
@@ -15,6 +15,7 @@ from lib389._constants import *
 from lib389.idm.user import UserAccounts
 from lib389.idm.organizationalunit import OrganizationalUnits
 from lib389.topologies import topology_st as topo
+from lib389.idm.directorymanager import DirectoryManager
 
 pytestmark = pytest.mark.tier1
 
@@ -25,14 +26,13 @@ else:
     logging.getLogger(__name__).setLevel(logging.INFO)
 log = logging.getLogger(__name__)
 
-USER_DN = 'uid=Test_user1,ou=People,dc=example,dc=com'
 USER_ACI = '(targetattr="userpassword")(version 3.0; acl "pwp test"; allow (all) userdn="ldap:///self";)'
-TOKEN = 'test_user1'
+TOKEN = 'test_user123'
 
 user_properties = {
-    'uid': 'Test_user1',
-    'cn': 'test_user1',
-    'sn': 'test_user1',
+    'uid': 'Test_user123',
+    'cn': 'test_user123',
+    'sn': 'test_user123',
     'uidNumber': '1001',
     'gidNumber': '2001',
     'userpassword': PASSWORD,
@@ -59,28 +59,48 @@ def test_token_lengths(topo):
     :id: dae9d916-2a03-4707-b454-9e901d295b13
     :setup: Standalone instance
     :steps:
-        1. Test token length rejects password of the same length as rdn value
+        1. Create user, setup global password policy
+        2. Bind as user, change password to 'Abcd012+'
+        3. Bind as user with 'Abcd012+', attempt changes to 'user', 'us123', 'Tuse!1234', 'Tuse!0987', 'Tabc!1234'
+        4. For each passwordMinTokenLength 4, 6, 10: change settings, rebind as user, attempt password with token of that length from TOKEN
+        5. Cleanup - delete user
     :expectedresults:
-        1. Passwords are rejected
+        1. User created, password policy enabled and set
+        2. Success
+        3. All attempts fail with CONSTRAINT_VIOLATION
+        4. All attempts fail with CONSTRAINT_VIOLATION
+        5. User successfully deleted
     """
     user = pwd_setup(topo)
-    for length in ['4', '6', '10']:
-        topo.standalone.simple_bind_s(DN_DM, PASSWORD)
-        topo.standalone.config.set('passwordMinTokenLength', length)
-        topo.standalone.simple_bind_s(USER_DN, PASSWORD)
-        time.sleep(1)
-
-        try:
-            passwd = TOKEN[:int(length)]
-            log.info("Testing password len {} token ({})".format(length, passwd))
-            user.replace('userpassword', passwd)
-            log.fatal('Password incorrectly allowed!')
-            assert False
-        except ldap.CONSTRAINT_VIOLATION as e:
-            log.info('Password correctly rejected: ' + str(e))
-        except ldap.LDAPError as e:
-            log.fatal('Unexpected failure ' + str(e))
-            assert False
+    dm = DirectoryManager(topo.standalone)
+
+    try:
+        # Verify that the user can change their password
+        user.rebind(PASSWORD)
+        user.replace('userpassword', 'Abcd012+')
+
+        # Verify that the default password policy is enforced
+        user.rebind('Abcd012+')
+        for new_password in ['user', 'us123', 'Tuse!1234', 'Tuse!0987', 'Tabc!1234']:
+            log.info(f"Testing password {new_password}")
+            with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+                user.replace('userpassword', new_password)
+
+        # Verify that the password policy is enforced for different token lengths
+        for length in ['4', '6', '10']:
+            dm.rebind(PASSWORD)
+            topo.standalone.config.set('passwordMinTokenLength', length)
+            user.rebind('Abcd012+')
+            time.sleep(1)
+
+            with pytest.raises(ldap.CONSTRAINT_VIOLATION):
+                passwd = TOKEN[:int(length)]
+                log.info("Testing password len {} token ({})".format(length, passwd))
+                user.replace('userpassword', passwd)
+
+    finally:
+        # Cleanup
+        user.delete()
 
 
 if __name__ == '__main__':
diff --git a/dirsrvtests/tests/tickets/ticket48896_test.py b/dirsrvtests/tests/tickets/ticket48896_test.py
deleted file mode 100644
index a1897589e..000000000
--- a/dirsrvtests/tests/tickets/ticket48896_test.py
+++ /dev/null
@@ -1,139 +0,0 @@
-# --- BEGIN COPYRIGHT BLOCK ---
-# Copyright (C) 2016 Red Hat, Inc.
-# All rights reserved.
-#
-# License: GPL (version 3 or any later version).
-# See LICENSE for details.
-# --- END COPYRIGHT BLOCK ---
-#
-import pytest
-from lib389.tasks import *
-from lib389.utils import *
-from lib389.topologies import topology_st
-
-from lib389._constants import DEFAULT_SUFFIX, DN_DM, PASSWORD
-
-# Skip on older versions
-pytestmark = [pytest.mark.tier2,
-              pytest.mark.skipif(ds_is_older('1.3.6'), reason="Not implemented")]
-
-logging.getLogger(__name__).setLevel(logging.DEBUG)
-log = logging.getLogger(__name__)
-
-CONFIG_DN = 'cn=config'
-UID = 'buser123'
-TESTDN = 'uid=%s,' % UID + DEFAULT_SUFFIX
-
-
-def check_attr_val(topology_st, dn, attr, expected):
-    try:
-        centry = topology_st.standalone.search_s(dn, ldap.SCOPE_BASE, 'cn=*')
-        if centry:
-            val = centry[0].getValue(attr)
-            if val == expected:
-                log.info('Default value of %s is %s' % (attr, expected))
-            else:
-                log.info('Default value of %s is not %s, but %s' % (attr, expected, val))
-                assert False
-        else:
-            log.fatal('Failed to get %s' % dn)
-            assert False
-    except ldap.LDAPError as e:
-        log.fatal('Failed to search ' + dn + ': ' + e.message['desc'])
-        assert False
-
-
-def replace_pw(server, curpw, newpw, expstr, rc):
-    log.info('Binding as {%s, %s}' % (TESTDN, curpw))
-    server.simple_bind_s(TESTDN, curpw)
-
-    hit = 0
-    log.info('Replacing password: %s -> %s, which should %s' % (curpw, newpw, expstr))
-    try:
-        server.modify_s(TESTDN, [(ldap.MOD_REPLACE, 'userPassword', ensure_bytes(newpw))])
-    except Exception as e:
-        log.info("Exception (expected): %s" % type(e).__name__)
-        hit = 1
-        assert isinstance(e, rc)
-
-    if (0 != rc) and (0 == hit):
-        log.info('Expected to fail with %s, but passed' % rc.__name__)
-        assert False
-
-    log.info('PASSED')
-
-
-def test_ticket48896(topology_st):
-    """
-    """
-    log.info('Testing Ticket 48896 - Default Setting for passwordMinTokenLength does not work')
-
-    log.info("Setting global password policy with password syntax.")
-    topology_st.standalone.simple_bind_s(DN_DM, PASSWORD)
-    topology_st.standalone.modify_s(CONFIG_DN, [(ldap.MOD_REPLACE, 'passwordCheckSyntax', b'on'),
-                                                (ldap.MOD_REPLACE, 'nsslapd-pwpolicy-local', b'on')])
-
-    config = topology_st.standalone.search_s(CONFIG_DN, ldap.SCOPE_BASE, 'cn=*')
-    mintokenlen = config[0].getValue('passwordMinTokenLength')
-    history = config[0].getValue('passwordInHistory')
-
-    log.info('Default passwordMinTokenLength == %s' % mintokenlen)
-    log.info('Default passwordInHistory == %s' % history)
-
-    log.info('Adding a user.')
-    curpw = 'password'
-    topology_st.standalone.add_s(Entry((TESTDN,
-                                        {'objectclass': "top person organizationalPerson inetOrgPerson".split(),
-                                         'cn': 'test user',
-                                         'sn': 'user',
-                                         'userPassword': curpw})))
-
-    newpw = 'Abcd012+'
-    exp = 'be ok'
-    rc = 0
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = 'user'
-    exp = 'fail'
-    rc = ldap.CONSTRAINT_VIOLATION
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = UID
-    exp = 'fail'
-    rc = ldap.CONSTRAINT_VIOLATION
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = 'Tuse!1234'
-    exp = 'fail'
-    rc = ldap.CONSTRAINT_VIOLATION
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = 'Tuse!0987'
-    exp = 'fail'
-    rc = ldap.CONSTRAINT_VIOLATION
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = 'Tabc!1234'
-    exp = 'fail'
-    rc = ldap.CONSTRAINT_VIOLATION
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    curpw = 'Abcd012+'
-    newpw = 'Direc+ory389'
-    exp = 'be ok'
-    rc = 0
-    replace_pw(topology_st.standalone, curpw, newpw, exp, rc)
-
-    log.info('SUCCESS')
-
-
-if __name__ == '__main__':
-    # Run isolated
-    # -s for DEBUG mode
-    CURRENT_FILE = os.path.realpath(__file__)
-    pytest.main("-s %s" % CURRENT_FILE)
-- 
2.52.0

